include/quda_matrix.h

```diff
@@ -1068,7 +1069,24 @@ namespace quda {
         - q(3) * q(1) * q(8) - q(0) * q(7) * q(5);
 
       Complex sg2h3 = sqrt(a3 * a3 - (Float)4. * a2 * a2 * a2);
-      Complex cp = exp(log((Float)0.5 * (a3 + sg2h3)) / (Float)3.0);
+
+      // If the matrix q is zero, this algorithm produces nan, instead of unity.
+      // The first invalid operation is the log hereafter. Therefore, we check
+      // if its argument (tmp) is zero and then return unity.
+
+      Complex tmp = a3 + sg2h3;
+      if (tmp.x == 0 and tmp.y==0) {
+	// Making sure q is a zero matrix
+	for(int i=0; i<9; i++) {
+	  assert(q(i).x==0 and q(i).y==0);
```

Comment:
I looked at the PTX/SASS for a smaller code tested in the compiler explorer, and this does generate a non-inlined function call. Whether or not it's harmless to end-to-end performance, I think it might be cleaner to not use `assert`.

Reply:
@weinbe2 thank you for having a look, then I would do the following:
- check if the matrix is actually zero and return identity
- otherwise let the algorithm go through and one will get `nan` in the output. This then should be a signal to the user that something wrong happened. I believe it is better `nan` than a wrong result.